src/main/java/com/github/luben/zstd/ZstdCompressCtx.java

```diff
@@ -268,6 +276,117 @@ public ZstdCompressCtx setLong(int windowLog) {
         return this;
     }
 
+    /**
+     * Register an external sequence producer
+     * @param producer the user-defined {@link SequenceProducer} to register.
+     */
+    public ZstdCompressCtx registerSequenceProducer(SequenceProducer producer) {
+        ensureOpen();
+        acquireSharedLock();
+        try {
+            if (this.seqprod != null) {
+                this.seqprod.freeState(seqprod_state);
+                this.seqprod = null;
+            }
+
+            if (producer == null) {
+                Zstd.registerSequenceProducer(nativePtr, 0, 0);
+            } else {
+                seqprod_state = producer.createState();
+                Zstd.registerSequenceProducer(nativePtr, seqprod_state, producer.getFunctionPointer());
+                this.seqprod = producer;
+            }
+        } catch (Exception e) {
+            this.seqprod = null;
+            Zstd.registerSequenceProducer(nativePtr, 0, 0);
+            releaseSharedLock();
```

Comment:
I think you don't need to release the lock here as it will be released in the `finally` block